fix(#4886) more: panics if file is not readable

Ludmuterol opened the pull request:
fix for https://github.com/uutils/coreutils/issues/4886#issue-1719861294
i just changed the unwrap to a match and i hope it actually is a UUsageError i am supposed to return here. I chose UsageError since the code above mine chose UsageError for a similar case

output of the preinstalled more on my linux: `more: cannot open foo: Permission denied`
output of this more: `./more: cannot open 'foo': permission denied 
Try './more --help' for more information.`

cakebaker commented:
Thanks for your PR :)

I wouldn't use a `UUsageError` in this case because the user didn't do anything wrong.

And can you please add a test to `tests/by-util/test_more.rs` to ensure we won't regress in the future?

Ludmuterol commented:
:) Sure! If I understand correctly i only need to add a file into the fixtures folder (in the correct sub folder) to be able to use it? i did so and changed it to the 244 perms and now every test panics...
```
failures:

---- test_more::test_more_dir_arg stdout ----
thread 'test_more::test_more_dir_arg' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 13, kind: PermissionDenied, message: "Permission denied" }', tests/common/util.rs:1134:76

---- test_more::test_valid_arg stdout ----
thread 'test_more::test_valid_arg' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 13, kind: PermissionDenied, message: "Permission denied" }', tests/common/util.rs:1134:76

---- test_more::test_more_no_arg stdout ----
thread 'test_more::test_more_no_arg' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 13, kind: PermissionDenied, message: "Permission denied" }', tests/common/util.rs:1134:76


failures:
    test_more::test_more_dir_arg
    test_more::test_more_no_arg
    test_more::test_valid_arg
```

i get that comes from recursive copy where it returns an error since it is not allowed to copy.
my idea would be to change the perms in the test, but then i'd need to do that OperatingSystem independent. std::fs::Permissions only supports readonly atm and we want to make it write only. I could go for a Unix solution: https://doc.rust-lang.org/std/os/unix/fs/trait.PermissionsExt.html but then that wont work on other OS's. any suggestions?

cakebaker commented:
You won't need fixtures in this case. And a Unix solution is fine. For inspiration you might want to have a look at the tests for `cp`: https://github.com/uutils/coreutils/blob/main/tests/by-util/test_cp.rs . Hope that helps :)

Ludmuterol commented:
![Screenshot from 2023-05-23 18-48-08](https://github.com/uutils/coreutils/assets/45458713/44573f7c-6370-411d-90bd-068168f48a54)

Changed to USimpleError and added a test case.
i have a weird newline(?) at the end of the output tho.. is that expected or an error on my part? (or even an error of my shell, etc..)

sylvestre commented:
please replace the screenshot by text: it is terrible for search and accessibility

sylvestre commented:
On windows, fails with:
```

error[E0433]: failed to resolve: could not find `unix` in `os`
 --> tests\by-util\test_more.rs:4:14
  |
4 | use std::os::unix::fs::PermissionsExt;
  |              ^^^^ could not find `unix` in `os`

warning: unused import: `std::io::Write`
 --> tests\by-util\test_du.rs:9:5
```

cakebaker commented:
> i have a weird newline(?) at the end of the output tho.. is that expected or an error on my part? (or even an error of my shell, etc..)
and i cant see any new characters i try to input unless i completely restart the shell. (although executing commands does work i only cant see them)

You forgot to disable "raw mode" before returning the error :)

> should i add a "skip test on not unix" inside my newly added test?

Yes, please, otherwise the CICD will fail when running it on Windows.

cakebaker commented on tests/by-util/test_more.rs:
```diff
@@ -32,3 +38,15 @@ fn test_more_dir_arg() {
+    if std::io::stdout().is_terminal() {
+        let (at, mut ucmd) = at_and_ucmd!();
+        let permissions = Permissions::from_mode(0o244);
+        at.make_file("invalid-perms.txt").metadata().unwrap();
```
Retrieving the metadata is unnecessary as we don't use them.
```suggestion
        at.make_file("invalid-perms.txt");
```

Ludmuterol commented:
Ah yes I missed that :) thank you!

cakebaker commented on tests/by-util/test_more.rs:
```diff
@@ -32,3 +38,15 @@ fn test_more_dir_arg() {
+        let permissions = Permissions::from_mode(0o244);
+        at.make_file("invalid-perms.txt").metadata().unwrap();
+        set_permissions(at.plus("invalid-perms.txt"), permissions).unwrap();
+        ucmd.arg("invalid-perms.txt").fails();
```
The `fails()` alone is not enough because currently the test passes even without applying your changes. I would enable the output check you had in a previous commit.

Ludmuterol replied:
ah yes looking at the cp tests i found stderr_contains. thats what i was missing :)

cakebaker commented on tests/by-util/test_more.rs:
```diff
@@ -1,6 +1,12 @@
 use crate::common::util::TestScenario;
 use is_terminal::IsTerminal;
 
+//Both following includes are only needed inside the test_more_invalid_file_perms()
+#[cfg(target_family = "unix")]
+use std::fs::{set_permissions, Permissions};
+#[cfg(target_family = "unix")]
+use std::os::unix::fs::PermissionsExt;
```
I would move the includes inside the function so you don't need the comment and the attributes.

cakebaker commented:
Thanks :)
